File: src/pipeline/test_detect.c
#include "foundation/constants.h"
#include "pipeline/pipeline.h"

#include <stdbool.h>
#include <stddef.h>
#include <string.h>

enum {
    PT_TEST_LEN = 4,
    PT_DESCRIBE_LEN = 9,
    PT_CONTEXT_LEN = 7,
};

#define SLEN(s) (sizeof(s) - 1)

#ifdef CBM_USE_RUST_PIPELINE_TEST_DETECT
extern int cbm_rs_pipeline_is_test_path_v1(const char *path);
extern int cbm_rs_pipeline_is_test_func_name_v1(const char *name);
#endif

static bool str_ends_with(const char *s, size_t slen, const char *suffix) {
    size_t sflen = strlen(suffix);
    return slen >= sflen && strcmp(s + slen - sflen, suffix) == 0;
}

bool cbm_is_test_path(const char *path) {
    if (!path) {
        return false;
    }
#ifdef CBM_USE_RUST_PIPELINE_TEST_DETECT
    return cbm_rs_pipeline_is_test_path_v1(path) != 0;
#endif
    const char *base = strrchr(path, '/');
    base = base ? base + SKIP_ONE : path;
    size_t len = strlen(path);

    if (strncmp(base, "test_", SLEN("test_")) == 0) {
        return true;
    }
    if (str_ends_with(path, len, "_test.go") || str_ends_with(path, len, "_test.py") ||
        str_ends_with(path, len, "_test.rs") || str_ends_with(path, len, "_test.cpp") ||
        str_ends_with(path, len, "_test.lua")) {
        return true;
    }
    if (strstr(path, ".test.ts") || strstr(path, ".spec.ts") || strstr(path, ".test.js") ||
        strstr(path, ".spec.js") || strstr(path, ".test.tsx") || strstr(path, ".spec.tsx")) {
        return true;
    }
    if (str_ends_with(path, len, "Test.java") || str_ends_with(path, len, "Test.kt") ||
        str_ends_with(path, len, "Test.cs") || str_ends_with(path, len, "Test.php") ||
        str_ends_with(path, len, "Spec.scala")) {
        return true;
    }
    if (strstr(path, "__tests__/") || strstr(path, "/tests/") || strstr(path, "/test/") ||
        strstr(path, "/spec/")) {
        return true;
    }
    if (strncmp(path, "tests/", SLEN("tests/")) == 0 ||
        strncmp(path, "test/", SLEN("test/")) == 0 || strncmp(path, "spec/", SLEN("spec/")) == 0 ||
        strncmp(path, "__tests__/", SLEN("__tests__/")) == 0) {
        return true;
    }
    return str_ends_with(path, len, "_spec.rb");
}

bool cbm_is_test_func_name(const char *name) {
    if (!name) {
        return false;
    }
#ifdef CBM_USE_RUST_PIPELINE_TEST_DETECT
    return cbm_rs_pipeline_is_test_func_name_v1(name) != 0;
#endif
    if (strncmp(name, "Test", SLEN("Test")) == 0 &&
        (name[PT_TEST_LEN] == '\0' || (name[PT_TEST_LEN] >= 'A' && name[PT_TEST_LEN] <= 'Z'))) {
        return true;
    }
    if (strncmp(name, "Benchmark", SLEN("Benchmark")) == 0 &&
        (name[PT_DESCRIBE_LEN] == '\0' ||
         (name[PT_DESCRIBE_LEN] >= 'A' && name[PT_DESCRIBE_LEN] <= 'Z'))) {
        return true;
    }
    if (strncmp(name, "Example", SLEN("Example")) == 0 &&
        (name[PT_CONTEXT_LEN] == '\0' ||
         (name[PT_CONTEXT_LEN] >= 'A' && name[PT_CONTEXT_LEN] <= 'Z'))) {
        return true;
    }
    if (strncmp(name, "test_", SLEN("test_")) == 0) {
        return true;
    }
    if (strncmp(name, "test", SLEN("test")) == 0 && name[PT_TEST_LEN] >= 'A' &&
        name[PT_TEST_LEN] <= 'Z') {
        return true;
    }
    if (strcmp(name, "test") == 0 || strcmp(name, "it") == 0 || strcmp(name, "describe") == 0 ||
        strcmp(name, "beforeAll") == 0 || strcmp(name, "afterAll") == 0 ||
        strcmp(name, "beforeEach") == 0 || strcmp(name, "afterEach") == 0) {
        return true;
    }
    return strcmp(name, "@testset") == 0 || strcmp(name, "@test") == 0;
}

File: rust/cbm-core/src/pipeline/test_detect.rs
//! Pipeline test file/function classifiers.
//!
//! These helpers mirror `src/pipeline/pass_tests.c` byte-for-byte enough for
//! opt-in dispatch. They do not create TESTS/TESTS_FILE edges or touch graph
//! state.

fn ends_with(bytes: &[u8], suffix: &[u8]) -> bool {
    bytes.len() >= suffix.len() && &bytes[bytes.len() - suffix.len()..] == suffix
}

fn starts_with(bytes: &[u8], prefix: &[u8]) -> bool {
    bytes.len() >= prefix.len() && &bytes[..prefix.len()] == prefix
}

fn contains(bytes: &[u8], needle: &[u8]) -> bool {
    !needle.is_empty() && bytes.windows(needle.len()).any(|window| window == needle)
}

#[must_use]
pub fn node_is_test(properties_json: Option<&[u8]>) -> bool {
    properties_json.is_some_and(|properties| contains(properties, b"\"is_test\":true"))
}

fn basename(path: &[u8]) -> &[u8] {
    match path.iter().rposition(|&byte| byte == b'/') {
        Some(pos) => &path[pos + 1..],
        None => path,
    }
}

fn find_subslice(haystack: &[u8], needle: &[u8]) -> Option<usize> {
    if needle.is_empty() {
        return Some(0);
    }
    haystack
        .windows(needle.len())
        .position(|window| window == needle)
}

fn build_path(path: &[u8], dir_len: usize, name: &[u8], ext: &[u8]) -> Vec<u8> {
    let mut out = Vec::with_capacity(dir_len + name.len() + ext.len() + usize::from(dir_len > 0));
    if dir_len > 0 {
        out.extend_from_slice(&path[..dir_len]);
        out.push(b'/');
    }
    out.extend_from_slice(name);
    out.extend_from_slice(ext);
    out
}

#[must_use]
pub fn is_test_path(path: Option<&[u8]>) -> bool {
    let Some(path) = path else {
        return false;
    };
    let base = basename(path);

    if starts_with(base, b"test_") {
        return true;
    }

    if ends_with(path, b"_test.go")
        || ends_with(path, b"_test.py")
        || ends_with(path, b"_test.rs")
        || ends_with(path, b"_test.cpp")
        || ends_with(path, b"_test.lua")
    {
        return true;
    }

    if contains(path, b".test.ts")
        || contains(path, b".spec.ts")
        || contains(path, b".test.js")
        || contains(path, b".spec.js")
        || contains(path, b".test.tsx")
        || contains(path, b".spec.tsx")
    {
        return true;
    }

    if ends_with(path, b"Test.java")
        || ends_with(path, b"Test.kt")
        || ends_with(path, b"Test.cs")
        || ends_with(path, b"Test.php")
        || ends_with(path, b"Spec.scala")
    {
        return true;
    }

    if contains(path, b"__tests__/")
        || contains(path, b"/tests/")
        || contains(path, b"/test/")
        || contains(path, b"/spec/")
    {
        return true;
    }

    if starts_with(path, b"tests/")
        || starts_with(path, b"test/")
        || starts_with(path, b"spec/")
        || starts_with(path, b"__tests__/")
    {
        return true;
    }

    ends_with(path, b"_spec.rb")
}

fn has_prefix_and_upper_or_end(name: &[u8], prefix: &[u8]) -> bool {
    starts_with(name, prefix)
        && (name.len() == prefix.len() || name[prefix.len()].is_ascii_uppercase())
}

#[must_use]
pub fn is_test_func_name(name: Option<&[u8]>) -> bool {
    let Some(name) = name else {
        return false;
    };

    if has_prefix_and_upper_or_end(name, b"Test")
        || has_prefix_and_upper_or_end(name, b"Benchmark")
        || has_prefix_and_upper_or_end(name, b"Example")
    {
        return true;
    }

    if starts_with(name, b"test_") {
        return true;
    }
    if starts_with(name, b"test") && name.len() > 4 && name[4].is_ascii_uppercase() {
        return true;
    }

    matches!(
        name,
        b"test"
            | b"it"
            | b"describe"
            | b"beforeAll"
            | b"afterAll"
            | b"beforeEach"
            | b"afterEach"
            | b"@testset"
            | b"@test"
    )
}

#[must_use]
pub fn test_to_prod_path(path: Option<&[u8]>) -> Option<Vec<u8>> {
    let path = path?;
    let base = basename(path);
    let dir_len = path.iter().rposition(|&byte| byte == b'/').unwrap_or(0);

    if ends_with(base, b"_test.go") {
        let stem_len = base.len() - b"_test.go".len();
        return Some(build_path(path, dir_len, &base[..stem_len], b".go"));
    }

    if starts_with(base, b"test_") && ends_with(base, b".py") {
        let stem = &base[b"test_".len()..base.len() - b".py".len()];
        return Some(build_path(path, dir_len, stem, b".py"));
    }

    for pat in [b".test.".as_slice(), b".spec.".as_slice()] {
        if let Some(pos) = find_subslice(base, pat) {
            return Some(build_path(path, dir_len, &base[..pos], &base[pos + 5..]));
        }
    }

    None
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn detects_test_paths_like_c_classifier() {
        for path in [
            b"foo_test.go".as_slice(),
            b"test_handler.py",
            b"handler.test.js",
            b"handler.spec.ts",
            b"Component.test.tsx",
            b"OrderTest.java",
            b"handler_test.rs",
            b"handler_test.cpp",
            b"OrderTest.cs",
            b"OrderTest.php",
            b"OrderSpec.scala",
            b"OrderTest.kt",
            b"handler_test.lua",
            b"__tests__/Component.js",
            b"tests/test_main.py",
            b"spec/handler_spec.rb",
            b"src/spec/handler.rb",
            b"src/__tests__/Component.js",
        ] {
            assert!(is_test_path(Some(path)), "{:?}", path);
        }

        for path in [
            b"foo.go".as_slice(),
            b"handler.py",
            b"handler.js",
            b"handler.ts",
            b"Component.tsx",
            b"Order.java",
            b"handler.rs",
            b"handler.cpp",
            b"Order.cs",
            b"Order.php",
            b"Order.scala",
            b"Order.kt",
            b"handler.lua",
            b"contest.go",
            b"latest.py",
        ] {
            assert!(!is_test_path(Some(path)), "{:?}", path);
        }
        assert!(!is_test_path(None));
    }

    #[test]
    fn detects_test_node_property_like_c_classifier() {
        assert!(node_is_test(Some(b"{\"is_test\":true}")));
        assert!(node_is_test(Some(b"{\"name\":\"x\",\"is_test\":true}")));
        assert!(!node_is_test(Some(b"{\"is_test\":false}")));
        assert!(!node_is_test(Some(b"{\"is_test\": true}")));
        assert!(!node_is_test(None));
    }

    #[test]
    fn mirrors_existing_path_edge_cases() {
        assert!(is_test_path(Some(b"handler.test.tsx")));
        assert!(is_test_path(Some(b"handler.spec.tsx")));
        assert!(is_test_path(Some(b"handler.test.ts.snap")));
        assert!(!is_test_path(Some(b"C:\\repo\\test_handler.py")));
        assert!(!is_test_path(Some(b"C:\\repo\\tests\\handler.py")));
        assert!(!is_test_path(Some(b"__tests__")));
        assert!(!is_test_path(Some(b"tests")));
        assert!(!is_test_path(Some(b"Test.java.bak")));
    }

    #[test]
    fn detects_test_function_names_like_c_classifier() {
        for name in [
            b"Test".as_slice(),
            b"TestCreate",
            b"TestHTTPHandler",
            b"Benchmark",
            b"BenchmarkParser",
            b"Example",
            b"ExampleReader",
            b"test_create",
            b"test",
            b"testCreate",
            b"describe",
            b"it",
            b"beforeAll",
            b"afterAll",
            b"beforeEach",
            b"afterEach",
            b"@testset",
            b"@test",
        ] {
            assert!(is_test_func_name(Some(name)), "{:?}", name);
        }

        for name in [
            b"create".as_slice(),
            b"handleRequest",
            b"process",
            b"Testable",
            b"Benchmarkable",
            b"Exampleable",
            b"testable",
            b"describeEach",
        ] {
            assert!(!is_test_func_name(Some(name)), "{:?}", name);
        }
        assert!(!is_test_func_name(None));
    }

    #[test]
    fn maps_test_paths_to_production_paths_like_c_contract() {
        assert_eq!(
            test_to_prod_path(Some(b"foo_test.go")).as_deref(),
            Some(b"foo.go".as_slice())
        );
        assert_eq!(
            test_to_prod_path(Some(b"nested/test_handler.py")).as_deref(),
            Some(b"nested/handler.py".as_slice())
        );
        assert_eq!(
            test_to_prod_path(Some(b"handler.test.ts")).as_deref(),
            Some(b"handler.ts".as_slice())
        );
        assert_eq!(
            test_to_prod_path(Some(b"handler.spec.tsx")).as_deref(),
            Some(b"handler.tsx".as_slice())
        );
        assert_eq!(
            test_to_prod_path(Some(b"handler.test.ts.bak")).as_deref(),
            Some(b"handler.ts.bak".as_slice())
        );
        assert_eq!(test_to_prod_path(Some(b"handler.go")), None);
        assert_eq!(test_to_prod_path(None), None);
    }
}
